### PR TITLE
Add ``end_lineno`` and ``end_col_offset`` to ``MessageLocationTuple``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Ref: #5336

Preparation for when `astroid` allows us to actually get these attributes. This allows us to start passing the needed attributes to reporters. Some points for consideration:

1) I have opted to type these as `Optional[int]`. I think it might be more valuable to know that a certain message does not have an associated `end_line` or `end_col_offset` instead of an arbitrarily chosen one. If we want to pass an arbitrary number we can still can, but at least now we don't need to do so.

2) I looked online but couldn't find an answer: is it possible to let users using `MessageLocationTuple` know that they should supply a `end_line` and `end_col_offset`? I think it is better to require users to think about whether they want `end_line` to be `None` at creation time (when they still have access to the `node`) then through the default value. Personally I would like these to be required in `pylint` `3.0`, but couldn't find a good way to deprecate `MessageLocationTuple` with 6 values.

3) The necessity of using default values for `MessageLocationTuple` also makes this require `python` > `3.6.0`. I'm not sure what the idea behind `2.12` was: do we allow > `3.6.0` functions in it? Or do we try to minimise them?

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks promising, I think there's also the functional tests to take into account and the output txt serialization/deserialization from testutil to upgrade ?  Also maybe it's time to introduce a location class with lineno, column, lineno_end, column_end attributes?

**Comment by DanielNoord** [comment]
> Looks promising, I think there's also the functional tests to take into account and the output txt serialization/deserialization from testutil to upgrade ?

I want to tackle those in another PR after we can actually access these attributes.

> Also maybe it's time to introduce a location class with lineno, column, lineno_end, column_end attributes?

That's what `MessageLocationTuple` is for :) We actually implemented that a couple of weeks ago!

**Comment by DanielNoord** [comment]
Should I fix the `TODO` in this PR and actually access `NodeNG.end_line` and `NodeNG.end_col_offset` or do we want to merge this first and keep the PRs small and manageable?

**Comment by Pierre-Sassoulas** [comment]
> I fix the TODO in this PR and actually access NodeNG.end_line and NodeNG.end_col_offset 

Yes, in order to avoid a double conflict when rebasing on master. We'll have to ask every contributor that touched the functional tests to rebase to get end_line / end_column for some time after merging this.

File: pylint/message/message.py [hunk 1]
```diff
@@ -25,6 +25,8 @@
         "obj",
         "line",
         "column",
+        "end_lineno",
+        "end_col_offset",
```
Comment (cdce8p) [inline pylint/message/message.py:29]: _Not a full review yet._

I was wondering, should we rename those to `end_line` and `end_column` respectively? That would match the existing `line` and `column`. Relevant as these are the arguments to `--msg-template`.

Comment (DanielNoord) [reply pylint/message/message.py:29]: I opted to follow what `ast` uses, but yeah this makes sense. I'll change!

**Comment by DanielNoord** [comment]
I'm not sure what tests we would like to add here. We actually only have one test for the `Message` class which gets updated in this PR.
Let me know if you want to add any additional.

I propose to merge this PR and then make three separate PRs, based on the original to-do list:
1) Modify reporters to accept `end_line` and `end_column`
2) Modify functional test runner to accept `end_line` and `end_column` with a `DeprecationWarning` if not supplied
3) Modify unittests to use `end_line` and `end_column`

Only 1 is necessary for the beta release of this functionality, although if we manage to do so I think it would be nice to put them all in the same release.

**Review by Pierre-Sassoulas — APPROVED** [review]
> I propose to merge this PR and then make three separate PRs, based on the original to-do list

👍

File: pylint/lint/pylinter.py [hunk 2]
```diff
@@ -1458,7 +1474,14 @@ def _add_one_message(
...
                 MessageLocationTuple(
-                    abspath, path, module or "", obj, line or 1, col_offset or 0
+                    abspath,
+                    path,
+                    module or "",
+                    obj,
+                    line or 1,
+                    col_offset or 0,
+                    end_lineno,
+                    end_col_offset,
```
Comment (cdce8p) [inline pylint/lint/pylinter.py:1484]: I was thinking again about this. Although I'm not 100% there yet, setting `end_lineno` and `end_col_offset` to `None` seems like the best we can do.

_We probably shouldn't touch `line or 1` and `col_offset or 0` though as this **will** break tools._

**Review by cdce8p — APPROVED** [review]
Ok lets try that! It at least seems to work with `vscode-python`. Although I'm still a bit unsure about the `None` values.

**Comment by Pierre-Sassoulas** [comment]
@DanielNoord I let you merge when it's ready :)

**Pull request merged**